### ISSUE TITLE
How to indicate that you can't test the truth value of an object?

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#5039]
This comes from #5024 

Consider this example:
```python
from typing_extensions import Never


class Foo:
    def __init__(self, x: int):
        self._x = x

    def __bool__(self) -> Never:
        raise Exception("bad idea")


y = Foo(10)
reveal_type(bool(y))

if y:
    print("should not happen")
```

pyright 1.1.305 reports the type of `bool(y)` as `bool`, but it should be `Never`, and then the `if` statement would be shown as being invalid.

So the question is how do we handle this with static type checking, i.e., make sure that you can't test the truth value of some class?  I thought overriding `__bool__()` returning `Never` or `NoReturn` would do the trick.

**Comment by erictraut** [comment]
The type of the expression `bool(y)` is dictated by the typeshed definition of the `bool` class constructor:

```python
class bool(int):
    def __new__(cls, __o: object = ...) -> Self: ...
```

If you want a static type checker to evaluate the type of the constructor otherwise, you (or the maintainers of typeshed) would need to update the `bool` class definition accordingly. For example:

```python
class NoSupportForBool(Protocol):
    def __bool__(self) -> NoReturn: ...

@final
class bool(int):
    @overload
    def __new__(cls, __o: NoSupportForBool = ...) -> NoReturn: ...
    @overload
    def __new__(cls, __o: object = ...) -> Self: ...
```

**Comment by Dr-Irv** [comment]
Even if that change is made, then pyright would not show that `if y` was invalid.  
```python
from typing import final, Protocol, overload
from typing_extensions import NoReturn
import boolop


class NoSupportForBool(Protocol):
    def __bool__(self) -> NoReturn:
        ...


@final
class bool(int):
    @overload
    def __new__(cls, __o: NoSupportForBool = ...) -> NoReturn:
        ...

    @overload
    def __new__(cls, __o: object = ...) -> boolop.bool:
        ...

    def __new__(cls, __o: object = ...) -> boolop.bool:
        return boolop.bool(__o)


class Foo(NoSupportForBool):
    def __init__(self, x: int):
        self._x = x

    def __bool__(self) -> NoReturn:
        raise Exception("bad idea")


y = Foo(10)

if y:
    print("should not happen")

reveal_type(bool(y))
```
The revealed type is `NoReturn`, but the `if` statement isn't shown as being invalid.

**Comment by erictraut** [comment]
The type of `y` is `Foo`, not `NoReturn`. If you use the expression `bool(y)` instead of `y` in the if statement, pyright will show that the statements below the `if` are unreachable.

![Screenshot 2023-04-30 at 9 54 04 AM](https://user-images.githubusercontent.com/7040122/235365812-f3bdd8c7-a278-46c8-9f2c-01c8966c09d6.png)

I think you're suggesting that pyright should model all conditional statements as if the `bool` constructor or the `__bool__` magic method is implicitly called on the operand. That's theoretically possible, but it would be very costly at code analysis time because this would need to be done in the hot path inside the code flow engine. This significant perf overhead would affect all users and all code (because conditional statements are very common) but would benefit very few users in a small number of edge-case situations. That's not a good tradeoff.

I think you're expecting too much from the type system here. There are many classes of bugs that cannot be caught using static typing. I think you'll need to accept that this is one such class.

**Comment by erictraut** [comment]
Thinking about this a bit more... 

A `NoReturn` (or `Never`) return type doesn't imply that a call to that function or method should be reported as a type violation. In other words, the call is not "invalid". It simply means that the function does not return control back to the caller via the normal return mechanism. That's why it affects control flow calculations.

Re-reading your initial post, it sounds like you are proposing that a `__bool__` method with a return type of `NoReturn` should imply that it is "invalid" to use that object as an operand to a conditional[^1]. Your proposal would require that type checkers apply a different (or additional) meaning to `NoReturn` in this particular case. I can't think of any other situations where `NoReturn` conferred the meaning "this is invalid". I seem to recall that the typing community has rejected other attempts to confer a broader meaning in cases of operator overloads like `__add__`, but I can't find those threads at the moment.

It might be reasonable in the particular case of the `__bool__` method to confer additional meaning to a `NoReturn` type. This interpretation would need to be agreed upon by the broader type checking community, however. I'm personally on the fence on this issue. I see the utility, but I'm a bit queasy about overloading the meaning of `NoReturn` in certain cases.

If you want to pursue this proposal further, I recommend starting a discussion thread in the [python/typing discussion forum](https://github.com/python/typing/discussions).

Note that my concerns about performance overhead are mitigated by this alternate interpretation of `NoReturn` for `__bool__`. Adding a separate "validity" check for conditional statements is a much lower overhead because it doesn't affect code flow analysis and therefore applies only to code that is being type checked. It doesn't affect imported source files whose types are being inferred. I'm therefore willing to revisit this issue and implement this additional check if there is agreement in the typing community.

[^1]: Conditionals include `if` statements, `while` statements, `assert` statements, an `if` clause in a ternary operator, a `not` operator, an `and` operator, an `or` operator, an `if` clause in a list comprehension, and an `if` clause in a `case` statement.

**Comment by gandhis1** [comment]
> A NoReturn (or Never) return type doesn't imply that a call to that function or method should be reported as a type violation.

I understand that a call shouldn't necessarily be reported as a type violation, but this is more than a call, this is an attempt to use the return value. Shouldn't that be a type violation?

**Comment by erictraut** [comment]
> ...this is an attempt to use the return value. Shouldn't that be a type violation?

`NoReturn` is compatible with all types in the Python type system. Calling a function that returns `NoReturn` cannot result in a type violation based on return type compatibility if the result is used in a covariant manner.

I think there's a good argument to be made that conditionals in Python don't use the result of `__bool__` in a covariant manner. They require a `bool` response and generate an exception if `__bool__` returns any other type. In other words, they treat the result as invariant, not covariant. That would imply that if `__bool__` is annotated with a return type of anything other than `bool`, it should be treated as an error when used in a conditional. That would eliminate any special-case interpretation for `NoReturn`.

**Comment by erictraut** [comment]
I think I'm convinced based on the above discussion. I've implemented this new check in pyright, and it will be included in the next release.

I think it's still important to discuss the issue in python/typing to get input from (and visibility with) the maintainers of typeshed and the other type checkers. If you want to rely on this check more generally, you'll need to get them to implement a similar check.

**Comment by Dr-Irv** [comment]
Discussion in python-typing:  https://github.com/python/typing/discussions/1398

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.306, which I just published. It will also be included in a future release of pylance.
